pkg/controller/direct/compute/firewallpolicyrule/firewallpolicyrule_controller.go

```diff
@@ -0,0 +1,297 @@
+	if err != nil {
+		return err
+	}
+
+	log := klog.FromContext(ctx).WithName(ctrlName)
+	log.V(2).Info("updating ComputeFirewallPolicyRule", "priority", a.priority)
+	mapCtx := &direct.MapContext{}
+
+	desired := a.desired.DeepCopy()
+	firewallPolicyRule := ComputeFirewallPolicyRuleSpec_ToProto(mapCtx, &desired.Spec)
+	if mapCtx.Err() != nil {
+		return mapCtx.Err()
+	}
+	// The field priority should be removed from the patch request body and included as a query parameter.
+	// See API doc: https://cloud.google.com/compute/docs/reference/rest/v1/firewallPolicies/patchRule#query-parameters
+	firewallPolicyRule.Priority = nil
+
+	updated := &computepb.FirewallPolicyRule{}
+
+	updateReq := &computepb.PatchRuleFirewallPolicyRequest{
+		FirewallPolicyRuleResource: firewallPolicyRule,
+		FirewallPolicy:             a.firewallPolicy,
+		Priority:                   direct.PtrTo(int32(a.priority)),
+	}
+	op, err := a.firewallPoliciesClient.PatchRule(ctx, updateReq)
+	if err != nil {
+		return fmt.Errorf("updating ComputeFirewallPolicyRule %d: %w", a.priority, err)
+	}
+	if !op.Done() {
+		err = op.Wait(ctx)
+		if err != nil {
+			return fmt.Errorf("waiting ComputeFirewallPolicyRule %d update failed: %w", a.priority, err)
+		}
+	}
+	log.V(2).Info("successfully updated ComputeFirewallPolicyRule", "priority", a.priority)
+
+	// Get the updated resource
+	updated, err = a.get(ctx)
+	if err != nil {
+		return fmt.Errorf("getting ComputeFirewallPolicyRule %d: %w", a.priority, err)
+	}
+
+	status := &krm.ComputeFirewallPolicyRuleStatus{}
+	status = ComputeFirewallPolicyRuleStatus_FromProto(mapCtx, updated)
+	return updateOp.UpdateStatus(ctx, status, nil)
+}
+
+func (a *firewallPolicyRuleAdapter) Export(ctx context.Context) (*unstructured.Unstructured, error) {
+	if a.actual == nil {
+		return nil, fmt.Errorf("firewallPolicyRule %d not found", a.priority)
+	}
+
+	mc := &direct.MapContext{}
+	spec := ComputeFirewallPolicyRuleSpec_FromProto(mc, a.actual)
+	specObj, err := runtime.DefaultUnstructuredConverter.ToUnstructured(spec)
+	if err != nil {
+		return nil, fmt.Errorf("error converting firewallPolicyRule spec to unstructured: %w", err)
+	}
+
+	u := &unstructured.Unstructured{
+		Object: make(map[string]interface{}),
+	}
+	u.SetGroupVersionKind(krm.ComputeFirewallPolicyRuleGVK)
+
+	if err := unstructured.SetNestedField(u.Object, specObj, "spec"); err != nil {
+		return nil, fmt.Errorf("setting spec: %w", err)
+	}
+
+	return u, nil
+}
+
+// Delete implements the Adapter interface.
+func (a *firewallPolicyRuleAdapter) Delete(ctx context.Context, deleteOp *directbase.DeleteOperation) (bool, error) {
+	log := klog.FromContext(ctx).WithName(ctrlName)
+	log.V(2).Info("deleting ComputeFirewallPolicyRule", "priority", a.priority)
+
+	delReq := &computepb.RemoveRuleFirewallPolicyRequest{
+		FirewallPolicy: a.firewallPolicy,
+		Priority:       direct.PtrTo(int32(a.priority)),
+	}
+	op, err := a.firewallPoliciesClient.RemoveRule(ctx, delReq)
+
```

Comment:
Nit: probably clearer to not have a blank line between the call and its error handling